HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift

```diff
@@ -15,45 +15,269 @@
 import XCTest
 @testable import HeartbeatLogging
 
-// TODO: Add additional validation (#8896 comments).
-
 class HeartbeatControllerTests: XCTestCase {
+  // 2021-11-01 @ 00:00:00 (EST)
+  let date = Date(timeIntervalSince1970: 1_635_739_200)
+
   func testFlushWhenEmpty() throws {
     // Given
     let controller = HeartbeatController(storage: HeartbeatStorageFake())
+    // Then
+    assertHeartbeatControllerFlushesEmptyPayload(controller)
+  }
+
+  func testLogAndFlush() throws {
+    // Given
+    let clock = SystemClock(date: date)
+
+    let controller = HeartbeatController(
+      storage: HeartbeatStorageFake(),
+      dateProvider: { clock.date }
+    )
+
+    assertHeartbeatControllerFlushesEmptyPayload(controller)
+
     // When
-    let flushed = controller.flush()
+    controller.log("dummy_agent")
+    let heartbeatPayload = controller.flush()
+
     // Then
-    XCTAssertNil(flushed)
+    assertEqualPayloadStrings(
+      heartbeatPayload.headerValue(),
+      """
+      {
+        "version": 0,
+        "payload": [
+          {
+            "agent": "dummy_agent",
+            "dates": ["2021-11-01"]
+          }
+        ]
+      }
+      """
+    )
+
+    assertHeartbeatControllerFlushesEmptyPayload(controller)
   }
 
-  func testLogThenFlush() throws {
+  func testLoggingDifferentAgentsInSameTimePeriodOnlyStoresTheFirst() throws {
     // Given
-    let controller = HeartbeatController(storage: HeartbeatStorageFake())
-    XCTAssertNil(controller.flush())
+    let clock = SystemClock(date: date)
+
+    let controller = HeartbeatController(
+      storage: HeartbeatStorageFake(),
+      dateProvider: { clock.date }
+    )
+
+    assertHeartbeatControllerFlushesEmptyPayload(controller)
+
+    // When
+    controller.log("dummy_agent")
+    controller.log("some_other_dummy_agent")
+    let heartbeatPayload = controller.flush()
+
+    // Then
+    assertEqualPayloadStrings(
+      heartbeatPayload.headerValue(),
+      """
+      {
+        "version": 0,
+        "payload": [
+          {
+            "agent": "dummy_agent",
+            "dates": ["2021-11-01"]
+          }
+        ]
+      }
+      """
+    )
+
+    assertHeartbeatControllerFlushesEmptyPayload(controller)
+  }
+
+  func testLogAtEndOfTimePeriodAndAcceptAtStartOfNextOne() throws {
+    // Given
+    let clock = SystemClock(date: date)
+
+    let controller = HeartbeatController(
+      storage: HeartbeatStorageFake(),
+      dateProvider: { clock.date }
+    )
+
+    assertHeartbeatControllerFlushesEmptyPayload(controller)
+
+    // When
+    // - Clock time 2021-11-01 @ 00:00:00 (EST)
+    controller.log("dummy_agent")
+
+    // - Advance to 2021-11-01 @ 23:59:59 (EST)
+    do { clock.advance(by: 60 * 60 * 24 - 1) }
```

Comment:
I'm curious what is the motivation to wrap this statement into `do {}`?

Reply:
I was experimenting with trying to differentiate public API usage from test logic to get us in the desired state.  I was trying to emphasize that the instructions in the `do {}` block are not public API related operations but test related code to get us in the desired state.

I also didnt know you could use `do {}` blocks like that. It makes for a nicer syntax then doing `_ = { ... }()`!

BUT, from looking at this again, I think renaming `clock`  to `testClock` makes it clear enough without silly `do {}` stuff.

Reply:
Reworked in e9d0361